pkg/sosreport/index.js

```diff
@@ -0,0 +1,180 @@
+        sos_archive_files = [ ];
+        sos_pid = null;
+
+        var task = cockpit.script('echo PID:$$; /usr/sbin/sosreport --batch', [ ],
+                                  { superuser: true, err: "out" });
+        sos_task = task;
+
+        // XXX - use a real API instead of scraping stdout
+        var output = "";
+        var pid_regex = /^PID:([0-9]+)\n/;
+        var progress_regex = /Running ([0-9]+)\/([0-9]+):/g;
+        var archive_regex = /Your sosreport has been generated and saved in:[ \n]+(\/[^\n]+)/;
+
+        task.stream(function (text) {
+            if (sos_task == task) {
+                var m, p;
+
+                output += text;
+
+                if (!sos_pid && (m = pid_regex.exec(output))) {
+                    sos_pid = m[1];
+                    output = output.slice(m[0].length);
+                }
+
+                p = 0;
+                while ((m = progress_regex.exec(output))) {
+                    p = (parseInt(m[1]) / parseInt(m[2])) * 100;
+                }
+                $("#sos-progress .progress-bar").css("width", p.toString()+"%");
+            }
+        });
+        task.done(function () {
+            if (sos_task == task) {
+                var m = archive_regex.exec(output);
+                if (m) {
+                    var archive = m[1];
+                    var basename = archive.replace(/.*\//, "");
+
+                    sos_archive_files = [ archive, archive + ".md5" ];
+
+                    var query = window.btoa(JSON.stringify({
+                        payload: "fsread1",
+                        binary: "raw",
+                        path: archive,
+                        superuser: true,
+                        external: {
+                            "content-disposition": 'attachement; filename="' + basename + '"',
+                            "content-type": "application/octet-stream"
+                        }
+                    }));
+                    sos_archive_url = "/cockpit/channel/" + cockpit.transport.csrf_token + '?' + query;
+                    $("#sos-progress, #sos-error").hide();
+                    $("#sos-alert, #sos-download").show();
+                    $("#sos-cancel").text(_("Close"));
+                } else {
+                    sos_error(_("No archive has been created."), output);
+                }
+                sos_task = null;
+            }
+        });
+        task.fail(function (error) {
+            if (sos_task == task) {
+                sos_error(error.toString(), output);
+                sos_task = null;
+            }
+        });
+    }
+
+    function sos_cancel() {
+        if (sos_task) {
+            if (sos_pid) {
+                /* HACK - https://bugzilla.redhat.com/show_bug.cgi?id=1278784
+                 * SIGTERM only aborts the current plugin, but SIGINT will stop the whole report
+                 */
+                cockpit.spawn([ "pkill", "-SIGINT", "-P", sos_pid ], { superuser: true, err: "message" }).
```

Comment:
Why do we use pkill here? This means yet another runtime dependency. A stream channel will kill its process if it receives close with a problem. Should we make that more explicit?